src/main/java/com/google/sps/servlets/BlobServlet.java

```diff
@@ -18,112 +22,374 @@
+ * allows creation and update of images and masks, and
+ * supports queries for images based on various parameters.
+ */
 @WebServlet("/blobs")
 public class BlobServlet extends HttpServlet {
 
+  /**
+   * Handles POST requests for images and masks.
+   * Responds with image URL and name upon successful POST.
+   * @param     {HttpServletRequest}    request
+   * @param     {HttpServletResponse}   response
+   * @return    {void}
+   */
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response)
       throws IOException {
     UserService userService = UserServiceFactory.getUserService();
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
 
+    // Must be logged in
     if (!userService.isUserLoggedIn()) {
-      response.sendRedirect("/imgupload.html");
+      response.sendRedirect("/");
       return;
     }
 
-    // Get the name entered by the user
-    String name = request.getParameter("name");
+    // Mode is a required parameter
+    boolean isCreateMode = DataUtils.parseMode(request, response);
+
+    // Check if working with image or mask
+    String parentImg = request.getParameter("parent-img");
+    boolean isMask = !DataUtils.isEmptyParameter(parentImg);
+
+    String now = Instant.now().toString();
+
+    // Get the image name entered by the user
+    // When creating, default image name if not provided is
+    // Untitled-{current UTC time}
+    String imgName = request.getParameter("img-name");
+    if (DataUtils.isEmptyParameter(imgName)) {
+      if (isCreateMode) {
+        imgName = "Untitled-" + now;
+      } else {
+        throw new IOException("Image name must be provided.");
+      }
+    }
+
+    String uEmail = userService.getCurrentUser().getEmail();
+    String projId = request.getParameter("proj-id");
+    Entity projEntity = DataUtils.getProjectEntity(projId, uEmail, true, false);
+    Key projKey = projEntity.getKey();
+    Key assetParentKey = projKey;
+
+    // Default case: creating a new base iamge
+    Entity imgEntity = new Entity(DataUtils.IMAGE, projKey);
+
+    // Asset to update must already exist
+    if (isMask) {
+      Entity existingImg = getAssetEntity(DataUtils.IMAGE, projKey, parentImg);
+
+      if (isCreateMode) {
+        Key parentEntityKey = existingImg.getKey();
+        imgEntity = new Entity(DataUtils.MASK, parentEntityKey);
+      } else {
+        assetParentKey = existingImg.getKey();
+        imgEntity = getAssetEntity(DataUtils.MASK, assetParentKey, imgName);
+      }
+    } else {
+      if (!isCreateMode) {
+        imgEntity = getAssetEntity(DataUtils.IMAGE, projKey, imgName);
+      }
+    }
+
+    // Owners have additional permissions
+    ArrayList<String> owners =
+        (ArrayList<String>)projEntity.getProperty("owners");
+    boolean isOwner = owners.contains(uEmail);
+
+    if (isCreateMode && !isOwner && !isMask) {
+      throw new IOException("You do not have permission to do that.");
```

Comment:
Let's be more specific than saying "to do that". Maybe "to upload an image to this project".